### PR TITLE
Allow providing a socketFactory on client creation

### PR DESCRIPTION
**Pull request by Nevon**
Allows for lower-level configuration of the socket connection by
injecting a custom factory on instantiation. Default implementation is
in src/network/socketFactory.js, which is automatically injected by the
entrypoint.

I haven't documented it yet, but the idea is that if you need access to the socket or want to provide a socket implementation other than the default one, you implement a function like `({  host, port, ssl, onConnect  }) => Socket`, where `Socket` is an instance of an object compatible with [`net.Socket`](https://nodejs.org/api/net.html#net_class_net_socket)

Fixes #258

Contributes to #36

@KBInmarsat FYI

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -17,152 +17,155 @@ const PRIVATE = {
   OFFSETS: Symbol('private:Kafka:offsets'),
 }
 
-module.exports = class Client {
```
Comment (Nevon) [inline src/index.js]: All of this is just prettier deciding

**Comment by tulios** [comment]
This PR is waiting for `v1.5.0` release

**Comment by Nevon** [comment]
I'm hoping to spend some time on Operation Chrome Trident either this week or the next, so it'd be nice to get a 👍 and get this puppy merged, @tulios

**Comment by tulios** [comment]
Yeah, I will merge this as soon as I release `1.5.1`

**Comment by KBInmarsat** [comment]
Thanks for this, I've been using it in our code for the past few days an it works fine.

**Pull request merged**